src/robot_localization_estimator.cpp

```diff
@@ -0,0 +1,330 @@
+{
+  // If newly received state is newer than any in the buffer, push back
+  if ( state_buffer_.empty() || state.time_stamp > state_buffer_.back().time_stamp )
+  {
+    state_buffer_.push_back(state);
+  }
+  // If it is older, put it in the right position
+  else
+  {
+    for ( boost::circular_buffer<EstimatorState>::iterator it = state_buffer_.begin(); it != state_buffer_.end(); ++it )
+    {
+      if ( state.time_stamp < it->time_stamp )
+      {
+        state_buffer_.insert(it, state);
+        return;
+      }
+    }
+  }
+}
+
+int RobotLocalizationEstimator::getState(const double time, EstimatorState& state) const
+{
+  // If there's nothing in the buffer, there's nothing to give.
+  if ( state_buffer_.size() == 0 )
+  {
+    return -1;
+  }
+  int return_value = 0;
+
+  // Set state to the most recent one for now
+  state = state_buffer_.back();
+
+  // Go through buffer from new to old
+  EstimatorState last_state_before_time = state_buffer_.front();
+  EstimatorState next_state_after_time = state_buffer_.back();
+  bool previous_state_found = false;
+  bool next_state_found = false;
+
+  for (boost::circular_buffer<EstimatorState>::const_reverse_iterator it = state_buffer_.rbegin(); it != state_buffer_.rend(); ++it)
+  {
+    /* If the time stamp of the current state from the buffer is
+       * older than the requested time, store it as the last state
+       * before the requested time. If it is younger, save it as the
+       * next one after, and go on to find the last one before.
+       */
+    if ( it->time_stamp <= time )
+    {
+      last_state_before_time = *it;
+      previous_state_found = true;
+      break;
+    }
+    else
+    {
+      next_state_after_time = *it;
+      next_state_found = true;
+    }
+  }
+
+  // If we found a previous state and a next state, we can do interpolation
+  if ( previous_state_found && next_state_found )
+  {
+    interpolate(last_state_before_time, next_state_after_time, time, state);
+  }
+
+  // If only a previous state is found, we can do extrapolation into the future
+  else if ( previous_state_found )
+  {
+    extrapolate(last_state_before_time, time, state);
+  }
+
+  // If only a next state is found, we'll have to extrapolate into the past.
+  else if ( next_state_found )
+  {
+    extrapolate(next_state_after_time, time, state);
+
+    // Warn the user that the buffer may be too small!
+    return_value = -2;
+  }
+
+  return return_value;
+}
+
+void RobotLocalizationEstimator::setBufferCapacity(const int capacity)
+{
+  state_buffer_.set_capacity(capacity);
+}
+
+void RobotLocalizationEstimator::clearBuffer()
+{
+  state_buffer_.clear();
+}
+
+unsigned int RobotLocalizationEstimator::capacity() const
+{
+  return state_buffer_.capacity();
+}
+
+unsigned int RobotLocalizationEstimator::size() const
+{
+  return state_buffer_.size();
+}
+
+void RobotLocalizationEstimator::extrapolate(const EstimatorState& boundary_state,
```

Comment:
This was fixed in 65d256cd4b297d430bb3c540a9eb907e3218d7bc and f24a843aad897e60282a1c3c4862af8743434bb9